compiler/noirc_evaluator/src/ssa/opt/pure.rs

```diff
@@ -0,0 +1,269 @@
+            let terminator = self.dfg[block].terminator();
+            if let Some(TerminatorInstruction::Return { return_values, .. }) = terminator {
+                if return_values.iter().any(&contains_reference) {
+                    return (Purity::Impure, BTreeSet::new());
+                }
+            }
+        }
+
+        (result, dependencies)
+    }
+}
+
+fn analyze_call_graph(
+    dependencies: HashMap<FunctionId, BTreeSet<FunctionId>>,
+    starting_purities: FunctionPurities,
+    main: FunctionId,
+) -> FunctionPurities {
+    let (graph, ids_to_indices, indices_to_ids) = build_call_graph(dependencies);
+
+    // Now we can analyze it: a function is only as pure as all of
+    // its called functions
+    let main_index = ids_to_indices[&main];
+    let mut dfs = Dfs::new(&graph, main_index);
+
+    // The `starting_purities` are the preliminary results from `is_pure`
+    // that don't take into account function calls. These finished purities do.
+    let mut finished_purities = HashMap::default();
+
+    while let Some(index) = dfs.next(&graph) {
+        let id = indices_to_ids[&index];
+        let mut purity = starting_purities[&id];
+
+        for neighbor_index in graph.neighbors(index) {
+            let neighbor = indices_to_ids[&neighbor_index];
+
+            let neighbor_purity = finished_purities.get(&neighbor).copied().unwrap_or({
+                // The dependent function isn't finished yet. Since we're following
+                // calls in a DFS, this means there are mutually recursive functions.
+                // We could handle these but would need a different, much slower algorithm
+                // to detect strongly connected components. Instead, since this should be
+                // a rare case, we bail and assume impure for now.
+                Purity::Impure
+            });
+            purity = purity.unify(neighbor_purity);
+        }
+
+        finished_purities.insert(id, purity);
+    }
+
+    finished_purities
+}
+
+fn build_call_graph(
+    dependencies: HashMap<FunctionId, BTreeSet<FunctionId>>,
+) -> (DiGraph<FunctionId, ()>, HashMap<FunctionId, PetGraphIndex>, HashMap<PetGraphIndex, FunctionId>)
+{
+    let mut graph = DiGraph::new();
+    let mut ids_to_indices = HashMap::default();
+    let mut indices_to_ids = HashMap::default();
+
+    for function in dependencies.keys() {
+        let index = graph.add_node(*function);
+        ids_to_indices.insert(*function, index);
+        indices_to_ids.insert(index, *function);
+    }
+
+    // Create edges from caller -> called
+    for (function, dependencies) in dependencies {
+        let function_index = ids_to_indices[&function];
+
+        for dependency in dependencies {
+            let dependency_index = ids_to_indices[&dependency];
+            graph.add_edge(function_index, dependency_index, ());
+        }
+    }
+
+    (graph, ids_to_indices, indices_to_ids)
+}
```

Comment:
We have some code in the inlining pass which builds up a callgraph so there's potential for code reuse here.

Reply:
I see compute_bottom_up_order but I'll leave them as separate for now since there's enough logic there having to do with weights that I think they are reasonably different. We probably should have a pass agnostic function to create a call graph eventually though. Perhaps this one can be reorganized in a later refactor.

Another thing I ran into was that I tried to account for mutually recursive functions in my first approach but would have needed a call graph which grouped SCCs into the same node which proved difficult. Having a call graph variant that did that (more expensive) as well would be nice.

Reply:
I'm also building a call graph in https://github.com/noir-lang/noir/pull/7188, but with slightly different data than in the inlining pass and this pass. I think it would be good to keep each PR as they are and then refactor later to create a pass agnostic call graph.